Sources/Auth/Core/AuthProvidable.swift

```diff
@@ -0,0 +1,18 @@
+//
+//  AuthProvidable.swift
+//  PovioKit
+//
+//  Created by Borut Tomazin on 20/01/2023.
+//  Copyright © 2023 Povio Inc. All rights reserved.
+//
+
+import PovioKitPromise
+import UIKit
+
+public protocol AuthProvidable {
+  typealias Authenticated = Bool
+  typealias Response = Authenticator.Response
+  
+  func signIn(from presentingViewController: UIViewController) -> Promise<Response>
```

Comment:
I think we can use [primary associated types](https://www.donnywals.com/what-are-primary-associated-types-in-swift-5-7/) to improve the design a bit.

We know that `Authenticator`s require `permissions` when logging in. Hence, it would make sense to add an associated type `Permission` to the protocol and change `login` function to accept a list of permissions.
But we know that this would lead to the old _protocol 'AuthProvidable can only be used as a generic constraint ..._ error.

Using primary associated types we can, I believe, avoid this issue. Here is how it would look in action:

```swift
public protocol AuthProvidable<Permission> {
    associatedtype Permission

    func signIn(
      permissions: [Permission]
      from presentingViewController: UIViewController
    ) -> Promise<Response>

    func signIn(
      from presentingViewController: UIViewController
    ) -> Promise<Response>
}

extension AuthProvidable {
    // implementations can provide their own methods, 
    // which can use _default permissions_
    func signIn(
      from presentingViewController: UIViewController
    ) -> Promise<Response> {
      signIn(permissions: [], from: presentingViewController)
    }
}
```

Specific protocols would look like this:

```swift
public protocol FacebookAuthProvidable: AuthProvidable<Permission> {
  var isAuthenticated: Authenticated { get }
}

public protocol GoogleAuthProvidable: AuthProvidable<()> {
  var isAuthenticated: Authenticated { get }
  static func shouldHandleURL(_ url: URL) -> Bool
}
```

---

What do you think?

Reply:
That's all good until it isn't. Every provider has it's own permissions to handle. This generic solution could solve this, but we'll leave that logic to the user to figure out the exact types for each platform. Which for me is not ideal.
I'll think about it

Reply:
Exactly. Each provider has its own permissions, which means that when you want to use a specific provider, you accept `any` implementation of `AuthProvidable`, which knows how to work with such permissions; i.e.:

```swift
func signInToFacebook(authenticator: any AuthProvidable<FacebookPermission>) -> ..
```